File: ios/SentryNativeContext.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// Builds the static-only `sentryContext` blob attached to the init
/// control frame. Backend's loader merges these onto every event via
/// `Sentry.addEventProcessor`, overwriting `nodeContextIntegration`'s
/// Darwin-libnode view with the actual user-facing iOS values.
///
/// Only fields that don't change during a session — battery, network,
/// foreground state etc. live with Phase 5's update frame.
enum SentryNativeContext {
    /// Returns the JSON-encoded blob ready to splice into the init
    /// frame. `nil` on failure so a builder bug never blocks boot.
    static func buildJSON() -> String? {
        guard let data = try? JSONSerialization.data(
            withJSONObject: build(),
            options: []
        ) else {
            return nil
        }
        return String(data: data, encoding: .utf8)
    }

    static func build() -> [String: Any] {
        let info = Bundle.main.infoDictionary ?? [:]
        let model = deviceModelIdentifier()
        let isSimulator: Bool = {
            #if targetEnvironment(simulator)
            return true
            #else
            return false
            #endif
        }()
        let family = familyForModel(model)
        let processorCount = ProcessInfo.processInfo.processorCount

        var device: [String: Any] = [
            "manufacturer": "Apple",
            "brand": "Apple",
            "model": model,
            "family": family,
            "arch": currentArch(),
            "simulator": isSimulator,
            "processor_count": processorCount,
            "memory_size": ProcessInfo.processInfo.physicalMemory,
        ]

        #if canImport(UIKit)
        let screen = UIScreen.main
        let bounds = screen.nativeBounds
        device["screen_resolution"] = "\(Int(bounds.width))x\(Int(bounds.height))"
        device["screen_density"] = screen.scale
        device["screen_dpi"] = Int(screen.scale * 160)
        #endif

        if let storage = totalStorageBytes() {
            device["storage_size"] = storage
        }

        var os: [String: Any] = [:]
        #if canImport(UIKit)
        os["name"] = UIDevice.current.systemName
        os["version"] = UIDevice.current.systemVersion
        #else
        os["name"] = "iOS"
        os["version"] = "unknown"
        #endif
        if let kernel = sysctlString("kern.osrelease") {
            os["kernel_version"] = kernel
        }
        if let build = sysctlString("kern.osversion") {
            os["build"] = build
        }

        var app: [String: Any] = [:]
        if let bundleId = Bundle.main.bundleIdentifier {
            app["app_identifier"] = bundleId
        }
        if let v = info["CFBundleShortVersionString"] as? String {
            app["app_version"] = v
        }
        if let b = info["CFBundleVersion"] as? String {
            app["app_build"] = b
        }
        if let n = info["CFBundleName"] as? String {
            app["app_name"] = n
        }

        let culture: [String: Any] = [
            "locale": Locale.current.identifier,
            "timezone": TimeZone.current.identifier,
        ]

        let tags: [String: Any] = [
            "os.name": (os["name"] as? String) ?? "iOS",
            "device.family": family,
            "device.simulator": String(isSimulator),
        ]

        return [
            "device": device,
            "os": os,
            "app": app,
            "culture": culture,
            "tags": tags,
        ]
    }

    /// e.g. "iPhone14,3", "iPad13,8". Matches what sentry-cocoa's
    /// `device.model` reports.
    private static func deviceModelIdentifier() -> String {
        var sysinfo = utsname()
        uname(&sysinfo)
        return withUnsafePointer(to: &sysinfo.machine) {
            $0.withMemoryRebound(to: CChar.self, capacity: 1) {
                String(cString: $0)
            }
        }
    }

    /// "iPhone" / "iPad" / "iPod" / "Mac" / "unknown".
    private static func familyForModel(_ model: String) -> String {
        if model.hasPrefix("iPhone") { return "iPhone" }
        if model.hasPrefix("iPad") { return "iPad" }
        if model.hasPrefix("iPod") { return "iPod" }
        if model.hasPrefix("Mac") || model.hasPrefix("arm") { return "Mac" }
        return "unknown"
    }

    private static func currentArch() -> String {
        #if arch(arm64)
        return "arm64"
        #elseif arch(x86_64)
        return "x86_64"
        #elseif arch(i386)
        return "i386"
        #else
        return "unknown"
        #endif
    }

    private static func sysctlString(_ name: String) -> String? {
        var size: size_t = 0
        sysctlbyname(name, nil, &size, nil, 0)
        guard size > 0 else { return nil }
        var buf = [CChar](repeating: 0, count: size)
        if sysctlbyname(name, &buf, &size, nil, 0) != 0 { return nil }
        return String(cString: buf)
    }

    private static func totalStorageBytes() -> Int64? {
        let url = URL(fileURLWithPath: NSHomeDirectory())
        let values = try? url.resourceValues(forKeys: [
            .volumeTotalCapacityKey,
        ])
        if let total = values?.volumeTotalCapacity { return Int64(total) }
        return nil
    }
}

File: android/src/main/java/com/comapeo/core/SentryNativeContext.kt
package com.comapeo.core

import android.content.Context
import android.content.res.Resources
import android.os.Build
import android.os.Environment
import android.os.StatFs
import androidx.core.content.pm.PackageInfoCompat
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonObject
import java.util.Locale
import java.util.TimeZone

/**
 * Builds the static-only `sentryContext` blob attached to the init
 * control frame. Backend's loader merges these onto every event via
 * `Sentry.addEventProcessor`, overwriting `nodeContextIntegration`'s
 * Linux-kernel view with the actual user-facing Android values.
 *
 * Only fields that don't change during a session — battery, network,
 * free memory, foreground state etc. live with Phase 5's update frame.
 */
object SentryNativeContext {
    fun build(context: Context): JsonObject {
        val pm = context.packageManager
        @Suppress("DEPRECATION")
        val pkg = if (Build.VERSION.SDK_INT >= 33) {
            pm.getPackageInfo(context.packageName, android.content.pm.PackageManager.PackageInfoFlags.of(0))
        } else {
            pm.getPackageInfo(context.packageName, 0)
        }
        val versionCode = PackageInfoCompat.getLongVersionCode(pkg)
        val metrics = Resources.getSystem().displayMetrics
        val storageStats = runCatching {
            val stat = StatFs(Environment.getDataDirectory().path)
            stat.blockSizeLong * stat.blockCountLong
        }.getOrNull()
        val totalMemoryBytes = runCatching {
            val mi = android.app.ActivityManager.MemoryInfo()
            (context.getSystemService(Context.ACTIVITY_SERVICE) as android.app.ActivityManager)
                .getMemoryInfo(mi)
            mi.totalMem
        }.getOrNull()
        val familyName = Build.PRODUCT
        // Generic / Android Studio AVD fingerprints contain "generic"
        // or "sdk_gphone"; ranger-1 era simulators use "ranchu". This
        // is best-effort — sentry-android does the same fingerprint
        // sniff for `device.simulator`.
        val isSimulator = Build.FINGERPRINT.startsWith("generic") ||
            Build.FINGERPRINT.contains("vbox") ||
            Build.MODEL.contains("Emulator") ||
            Build.MODEL.contains("Android SDK")

        return buildJsonObject {
            putJsonObject("device") {
                put("manufacturer", Build.MANUFACTURER)
                put("brand", Build.BRAND)
                put("model", Build.MODEL)
                put("model_id", Build.DEVICE)
                put("family", familyName)
                put("arch", Build.SUPPORTED_ABIS.firstOrNull() ?: "unknown")
                put("simulator", isSimulator)
                put("processor_count", Runtime.getRuntime().availableProcessors())
                put(
                    "screen_resolution",
                    "${metrics.widthPixels}x${metrics.heightPixels}",
                )
                put("screen_density", metrics.density.toDouble())
                put("screen_dpi", metrics.densityDpi)
                if (totalMemoryBytes != null) put("memory_size", totalMemoryBytes)
                if (storageStats != null) put("storage_size", storageStats)
            }
            putJsonObject("os") {
                put("name", "Android")
                put("version", Build.VERSION.RELEASE)
                put("build", Build.DISPLAY)
                put("kernel_version", System.getProperty("os.version") ?: "unknown")
            }
            putJsonObject("app") {
                put("app_identifier", context.packageName)
                pkg.versionName?.let { put("app_version", it) }
                put("app_build", versionCode.toString())
                // applicationInfo is nullable on API 33+; skip the
                // label rather than risk an NPE on a malformed pkg.
                pkg.applicationInfo?.let {
                    put("app_name", pm.getApplicationLabel(it).toString())
                }
            }
            putJsonObject("culture") {
                put("locale", Locale.getDefault().toLanguageTag())
                put("timezone", TimeZone.getDefault().id)
            }
            putJsonObject("tags") {
                put("os.name", "Android")
                put("device.family", familyName)
                put("device.simulator", isSimulator.toString())
            }
        }
    }
}
